### PR TITLE
migrate to zap logging

### PR DESCRIPTION
**Pull request by ewollesen**
As requested originally in #126 then re-worked to build on top of #134 

BACK-2500

### CODE REVIEW

**Comment by ewollesen** [comment]
TL;DR If you see a logging site that has lost context, **please point it out specifically!**

Included in this new PR is another round of review for every logging site, looking at the potential loss of any context. Sometimes it may look like a loss of context at first glance, but there are several factors that aren't always obvious that I took into consideration. A partial list:

1. if the function caller logs the same information
2. if the information is now included via the context logger (e.g. userID or the function caller's name)
3. some calls were demoted from warn to info, as warn will include a backtrace, and where there's no error to report on, backtraces don't make much sense

So please, if you see a place where you think context has been lost, please double check that one of the above doesn't replace the missing context/information, and if there's still a lack somewhere, then please call out each site. I've been through every logging site laboriously looking for places where I missed something, and I'd really rather not having to comb over everything yet again. :) Thank you!

File: api/clinic.go [hunk 1]
```diff
@@ -38,19 +38,17 @@ func (a *Api) InviteClinic(res http.ResponseWriter, req *http.Request, vars map[
...
 		} else if permissions["root"] == nil && permissions["custodian"] == nil {
-			a.sendError(res, http.StatusUnauthorized, STATUS_UNAUTHORIZED)
+			a.sendError(ctx, res, http.StatusUnauthorized, STATUS_UNAUTHORIZED)
 			return
 		}
 
 		defer req.Body.Close()
 		var ib = &ClinicInvite{}
 		if err := json.NewDecoder(req.Body).Decode(ib); err != nil {
-			a.logger.Errorw("error decoding invite", zap.Error(err))
```
Comment (ewollesen) [inline api/clinic.go:51]: All of this info is included in the `sendError` call.

File: api/forgot.go [hunk 2]
```diff
@@ -85,37 +86,27 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
 			a.logMetricAsServer("reset confirmation sent")
 		} else {
 			a.logMetricAsServer("reset confirmation failed to be sent")
-			log.Print("Something happened generating a passwordReset email")
```
Comment (ewollesen) [inline api/forgot.go:88]: I'm leaving this out as a) `addOrUpdateConfirmation` already does logging, and b) it's not useful.

File: api/forgot.go [hunk 3]
```diff
@@ -85,37 +86,27 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
...
 // find the reset confirmation if it exists and hasn't expired
-func (a *Api) findResetConfirmation(conf *models.Confirmation, ctx context.Context, res http.ResponseWriter) *models.Confirmation {
-
-	log.Printf("findResetConfirmation: finding [%v]", conf)
-	found, err := a.findExistingConfirmation(ctx, conf, res)
+func (a *Api) findResetConfirmation(ctx context.Context, conf *models.Confirmation) (*models.Confirmation, bool, error) {
+	a.logger(ctx).With("conf", conf).Debug("finding reset confirmation")
+	found, err := a.Store.FindConfirmation(ctx, conf)
 	if err != nil {
-		log.Printf("findResetConfirmation: error [%s]\n", err.Error())
```
Comment (ewollesen) [inline api/forgot.go:101]: The caller of this function will report the error, there' no need for duplications.

The same applies to all the logging in this function. The context of whether it wasn't found, or expired is maintained. 

Here's the only place this method is called if you want to double check:

https://github.com/tidepool-org/hydrophone/blob/eric%2Fsend-invite-v2.2-logging/api/forgot.go#L138

File: api/hydrophoneApi.go [hunk 4]
```diff
@@ -272,32 +346,17 @@ func (a *Api) IsAlive(res http.ResponseWriter, req *http.Request) {
...
-		return found, nil
-	}
-}
-
 // Find this confirmation
 // write error if it fails
 func (a *Api) addProfile(conf *models.Confirmation) error {
 	if conf.CreatorId != "" {
 		if err := a.seagull.GetCollection(conf.CreatorId, "profile", a.sl.TokenProvide(), &conf.Creator.Profile); err != nil {
-			log.Printf("error getting the creators profile [%v] ", err)
```
Comment (ewollesen) [inline api/hydrophoneApi.go:300]: I double checked that the caller logs the error in all cases.

File: api/invite.go [hunk 5]
```diff
@@ -168,54 +181,44 @@ func (a *Api) GetSentInvitations(res http.ResponseWriter, req *http.Request, var
...
+		conf, err := a.Store.FindConfirmation(ctx, accept)
 		if err != nil {
-			log.Printf("AcceptInvite error while finding confirmation [%s]\n", err.Error())
-			a.sendModelAsResWithStatus(res, err, http.StatusInternalServerError)
+			a.sendError(ctx, res, http.StatusInternalServerError, STATUS_ERR_FINDING_CONFIRMATION, err)
 			return
 		}
 		if conf == nil {
-			statusErr := &status.StatusError{Status: status.NewStatus(http.StatusNotFound, statusInviteNotFoundMessage)}
-			log.Println("AcceptInvite ", statusErr.Error())
```
Comment (ewollesen) [inline api/invite.go:217]: AcceptInvite will be logged by zap, as it logs the function name from which it was called.

File: api/invite.go [hunk 6]
```diff
@@ -227,61 +230,37 @@ func (a *Api) AcceptInvite(res http.ResponseWriter, req *http.Request, vars map[
 			ValidateCreatorID(invitorID, &validationErrors)
 
 		if len(validationErrors) > 0 {
-			for _, validationError := range validationErrors {
-				log.Println("AcceptInvite forbidden as there was a expectation mismatch", validationError)
```
Comment (ewollesen) [inline api/invite.go:231]: I feel like the existence of validation errors conveys the expectation mismatch, so I've removed that bit of extra text.

File: api/invite.go [hunk 7]
```diff
@@ -227,61 +230,37 @@ func (a *Api) AcceptInvite(res http.ResponseWriter, req *http.Request, vars map[
...
 		ctc := &models.CareTeamContext{}
 		if err := conf.DecodeContext(ctc); err != nil {
-			statusErr := &status.StatusError{Status: status.NewStatus(http.StatusBadRequest, STATUS_ERR_DECODING_CONTEXT)}
-			a.sendModelAsResWithStatus(res, statusErr, http.StatusBadRequest)
+			a.sendError(ctx, res, http.StatusBadRequest, STATUS_ERR_DECODING_CONTEXT)
 			return
 		}
 
 		if err := ctc.Validate(); err != nil {
-			log.Printf("AcceptInvite error validating CareTeamContext: %s", err)
```
Comment (ewollesen) [inline api/invite.go:249]: This will all be included between the caller name functionality of zap and the backtrace that zap will include.

File: api/invite.go [hunk 8]
```diff
@@ -322,23 +301,22 @@ func (a *Api) CancelInvite(res http.ResponseWriter, req *http.Request, vars map[
...
-			if a.addOrUpdateConfirmation(req.Context(), conf, res) {
+			// addOrUpdateConfirmation logs and writes a response on errors
+			if a.addOrUpdateConfirmation(ctx, conf, res) {
 				a.logMetric("canceled invite", req)
 				res.WriteHeader(http.StatusOK)
 				return
 			}
 		}
-		statusErr := &status.StatusError{Status: status.NewStatus(http.StatusNotFound, statusInviteNotFoundMessage)}
-		log.Printf("CancelInvite: [%s]", statusErr.Error())
```
Comment (ewollesen) [inline api/invite.go:340]: This is of course all included with the call to `sendError`.

File: api/invite.go [hunk 9]
```diff
@@ -460,22 +428,22 @@ func (a *Api) SendInvite(res http.ResponseWriter, req *http.Request, vars map[st
...
-			a.logger.Infof("invited [%s] user is already a member of the care team of %v", ib.Email, invitorID)
-			statusErr := &status.StatusError{Status: status.NewStatus(http.StatusConflict, statusExistingMemberMessage)}
-			a.sendModelAsResWithStatus(res, statusErr, http.StatusConflict)
+			a.sendError(ctx, res, http.StatusConflict, statusExistingMemberMessage,
+				zap.String("email", ib.Email), zap.String("invitorID", invitorID))
 			return
 		}
+
 		for key := range perms {
-			log.Printf("adding permission: %q %+v", key, perms[key])
```
Comment (ewollesen) [inline api/invite.go:476]: I'm collapsing these multiple log messages into a single message that captures all the permissions.

File: api/patientInvites.go [hunk 10]
```diff
@@ -138,24 +127,21 @@ func (a *Api) CancelOrDismissPatientInvite(res http.ResponseWriter, req *http.Re
...
+			a.sendError(ctx, res, http.StatusForbidden, statusInviteNotFoundMessage)
 			return
 		}
 
 		updatedStatus := models.StatusCanceled
 		if token.UserID != conf.CreatorId {
 			updatedStatus = models.StatusDeclined
+			// assertClinicMember logs and writes a response on errors
 			if err := a.assertClinicMember(ctx, clinicId, token, res); err != nil {
-				a.logger.Errorw("token owner is not a clinic member", zap.Error(err))
```
Comment (ewollesen) [inline api/patientInvites.go:158]: In this case, the message is misleading. A network outage could have caused the error. There's no way for this function to know this kind of detail.

In any event, `assertClinicMember` does its own logging, including a backtrace, so whatever the cause is, it should be apparent.

File: api/signup.go [hunk 11]
```diff
@@ -4,59 +4,29 @@ import (
...
 	clinics "github.com/tidepool-org/clinic/client"
 
 	commonClients "github.com/tidepool-org/go-common/clients"
 	"github.com/tidepool-org/go-common/clients/shoreline"
-	"github.com/tidepool-org/go-common/clients/status"
 
 	"github.com/tidepool-org/hydrophone/models"
 )
 
-const (
```
Comment (ewollesen) [inline api/signup.go:21]: These constants were all moved into one file for centralization, and a few duplicates were removed.

File: api/signup.go [hunk 12]
```diff
@@ -65,40 +35,36 @@ func (a *Api) findSignUp(ctx context.Context, conf *models.Confirmation, res htt
...
 	if err := json.NewDecoder(req.Body).Decode(fromBody); err != nil {
-		log.Printf("updateSignupConfirmation: error decoding signup to cancel [%s]", err.Error())
-		statusErr := &status.StatusError{Status: status.NewStatus(http.StatusBadRequest, STATUS_ERR_DECODING_CONFIRMATION)}
-		a.sendModelAsResWithStatus(res, statusErr, http.StatusBadRequest)
+		a.sendError(ctx, res, http.StatusBadRequest, STATUS_ERR_DECODING_CONFIRMATION, err)
 		return
 	}
 
 	if fromBody.Key == "" {
-		log.Printf("updateSignupConfirmation: %s", STATUS_SIGNUP_NO_CONF)
```
Comment (ewollesen) [inline api/signup.go:77]: Reminder: zap will include the calling function name.

File: api/signup.go [hunk 13]
```diff
@@ -163,7 +132,6 @@ func (a *Api) sendSignUp(res http.ResponseWriter, req *http.Request, vars map[st
 			return
 		} else {
 			a.logMetric("signup confirmation failed to be sent", req)
-			log.Print("Something happened generating a signup email")
```
Comment (ewollesen) [inline api/signup.go:166]: Not useful. Plus, `createAndSendNotification` writes log messages that actually contain useful information.

File: api/signup.go [hunk 14]
```diff
@@ -232,7 +202,6 @@ func (a *Api) resendSignUp(res http.ResponseWriter, req *http.Request, vars map[
 					a.logMetricAsServer("signup confirmation re-sent")
 				} else {
 					a.logMetricAsServer("signup confirmation failed to be sent")
-					log.Print("resendSignUp: Something happened trying to resend a signup email")
```
Comment (ewollesen) [inline api/signup.go:235]: Not useful. Plus, `createAndSendNotification` writes log messages that actually contain useful information.

File: clients/mongoStoreClient.go [hunk 15]
```diff
@@ -144,7 +145,6 @@ func (c *MongoStoreClient) FindConfirmation(ctx context.Context, confirmation *m
 	opts := options.FindOne().SetSort(bson.D{{Key: "created", Value: -1}})
 
 	if err = confirmationsCollection(c).FindOne(ctx, query, opts).Decode(&result); err != nil && err != mongo.ErrNoDocuments {
-		log.Printf("FindConfirmation: something bad happened [%v]", err)
```
Comment (ewollesen) [inline clients/mongoStoreClient.go:147]: I double checked that the caller logs these errors.

File: clients/sesNotifier.go [hunk 16]
```diff
@@ -110,23 +110,7 @@ func (c *SesNotifier) Send(to []string, subject string, msg string) (int, string
 
 	// Display error messages if they occur.
 	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
```
Comment (ewollesen) [inline clients/sesNotifier.go:113]: All of this information is output in the string representation of the AWS error, so nothing's lost here.

File: events/events.go [hunk 17]
```diff
@@ -26,7 +25,6 @@ func NewHandler(store clients.StoreClient) events.EventHandler {
 }
 
 func (h *handler) HandleDeleteUserEvent(payload events.DeleteUserEvent) error {
-	log.Printf("Deleting confirmations for user %v", payload.UserID)
```
Comment (ewollesen) [inline events/events.go:29]: Removed as not useful. I guess I could re-add it as a trace-level, if this is somehow useful to someone.

File: api/forgot.go [hunk 18]
```diff
@@ -133,46 +124,50 @@ func (a *Api) findResetConfirmation(conf *models.Confirmation, ctx context.Conte
...
-				a.sendModelAsResWithStatus(
-					res,
-					status.StatusError{Status: status.NewStatus(http.StatusOK, STATUS_RESET_ACCEPTED)},
-					http.StatusOK,
-				)
-				return
-			}
+		conf.UpdateStatus(models.StatusCompleted)
+		// addOrUpdateConfirmation logs and writes a response on errors
+		if !a.addOrUpdateConfirmation(ctx, conf, res) {
```
[the remark this answers is not in the document]
Comment (toddkazakov) [reply api/forgot.go:167]: I just double checked. This is definitely incorrect.

File: api/patientInvites.go [hunk 19]
```diff
@@ -85,38 +86,26 @@ func (a *Api) AcceptPatientInvite(res http.ResponseWriter, req *http.Request, va
...
 		}
 
 		conf.UpdateStatus(models.StatusCompleted)
-		if !a.addOrUpdateConfirmation(req.Context(), conf, res) {
-			a.logger.Warn("error adding or updating confirmation")
-			statusErr := &status.StatusError{Status: status.NewStatus(http.StatusInternalServerError, STATUS_ERR_SAVING_CONFIRMATION)}
-			a.sendModelAsResWithStatus(res, statusErr, http.StatusInternalServerError)
+		// addOrUpdateConfirmation logs and writes a response on errors
+		if !a.addOrUpdateConfirmation(ctx, conf, res) {
+			a.sendError(ctx, res, http.StatusInternalServerError, STATUS_ERR_SAVING_CONFIRMATION, err)
```
Comment (toddkazakov) [inline api/patientInvites.go:103]: Error is already sent at this point

**Review by toddkazakov — CHANGES_REQUESTED** [review]
Found some issues and some nitpicks

File: events/events.go (hunk 17 above)
Comment (ewollesen) [reply events/events.go:29]: Will fix.

File: api/clinic.go [hunk 20]
```diff
@@ -79,26 +77,24 @@ func (a *Api) InviteClinic(res http.ResponseWriter, req *http.Request, vars map[
...
+			a.sendError(ctx, res, http.StatusConflict, statusExistingPatientMessage,
+				"user is already a patient of clinic")
 			return
 		}
-		existingInvite, err := a.checkForDuplicateClinicInvite(req.Context(), clinicId, inviterID)
+		existingInvite, err := a.checkForDuplicateClinicInvite(ctx, clinicId, inviterID)
 		if err != nil {
-			a.sendError(res, http.StatusInternalServerError, STATUS_ERR_FINDING_CONFIRMATION, err)
+			a.sendError(ctx, res, http.StatusInternalServerError, STATUS_ERR_FINDING_CONFIRMATION, err,
+				fmt.Sprintf("clinic %s user already has or had an invite from %v", clinicId, inviterID))
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/clinic.go:92]: OK. I just copied the existing text, but I can do that.

File: api/clinicianInvites.go [hunk 21]
```diff
@@ -78,9 +75,9 @@ func (a *Api) SendClinicianInvite(res http.ResponseWriter, req *http.Request, va
 			return
 		}
 
-		statusErr := a.sendClinicianConfirmation(req, confirmation)
-		if statusErr != nil {
-			a.sendError(res, statusErr.Code, statusErr.Reason, statusErr.Error())
+		code, msg, optionalErr := a.sendClinicianConfirmation(req, confirmation)
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/clinicianInvites.go:78]: Turns out the code is superfluous anyway, removing it.

File: api/clinicianInvites.go [hunk 22]
```diff
@@ -402,31 +394,26 @@ func (a *Api) sendClinicianConfirmation(req *http.Request, confirmation *models.
...
 	if !a.createAndSendNotification(req, confirmation, emailContent) {
-		return &status.StatusError{
-			Status: status.NewStatus(http.StatusInternalServerError, STATUS_ERR_SENDING_EMAIL),
-		}
+		return http.StatusInternalServerError, STATUS_ERR_SENDING_EMAIL, nil
 	}
 
 	a.logMetric("clinician_invite_sent", req)
-	return nil
+	return 0, "", nil
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/clinicianInvites.go:401]: Removed.

File: api/forgot.go [hunk 23]
```diff
@@ -51,29 +51,30 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
...
 		if err != nil {
-			a.sendError(res, http.StatusInternalServerError, STATUS_ERR_CREATING_CONFIRMATION, err)
+			a.sendError(ctx, res, http.StatusInternalServerError, STATUS_ERR_CREATING_CONFIRMATION, err)
 			return
 		}
 
 		resetCnf.Email = email
 		//there is nothing more to do other than notify the user
 		resetCnf.UpdateStatus(models.StatusCompleted)
+		a.logger(ctx).With(zap.String("email", email)).Info(STATUS_RESET_NO_ACCOUNT)
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/forgot.go:73]: After reviewing what was there previously, I've removed this line.

File: api/hydrophoneApi.go [hunk 24]
```diff
@@ -148,103 +218,108 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
...
+	send.Handle("/forgot/{useremail}", vars(a.passwordReset)).Methods("POST")
+	send.Handle("/invite/{userid}", uid(vars(a.SendInvite))).Methods("POST")
+	send.Handle("/invite/{userId}/clinic", uid(vars(a.InviteClinic))).Methods("POST")
 
 	// POST /confirm/resend/signup/:useremail
 	// POST /confirm/resend/invite/:inviteId
-	c.Handle("/resend/signup/{useremail}", varsHandler(a.resendSignUp)).Methods("POST")
-	c.Handle("/resend/invite/{inviteId}", varsHandler(a.ResendInvite)).Methods("PATCH")
+	c.Handle("/resend/signup/{useremail}", vars(a.resendSignUp)).Methods("POST")
+	c.Handle("/resend/invite/{inviteId}", uid(vars(a.ResendInvite))).Methods("PATCH")
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/hydrophoneApi.go:244]: ...and I double checked the rest.

File: api/hydrophoneApi.go [hunk 25]
```diff
@@ -148,103 +218,108 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
...
-	c.Handle("/v1/clinics/{clinicId}/invites/clinicians", varsHandler(a.SendClinicianInvite)).Methods("POST")
-	c.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.ResendClinicianInvite)).Methods("PATCH")
-	c.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.GetClinicianInvite)).Methods("GET")
-	c.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.CancelClinicianInvite)).Methods("DELETE")
-
-	rtr.Handle("/v1/clinics/{clinicId}/invites/clinicians", varsHandler(a.SendClinicianInvite)).Methods("POST")
-	rtr.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.GetClinicianInvite)).Methods("GET")
-	rtr.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.ResendClinicianInvite)).Methods("PATCH")
-	rtr.Handle("/v1/clinics/{clinicId}/invites/clinicians/{inviteId}", varsHandler(a.CancelClinicianInvite)).Methods("DELETE")
+	c.Handle("/v1/clinics/{clinicId}/invites/patients", vars(a.GetPatientInvites)).Methods("GET")
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/hydrophoneApi.go:298]: Will take a stab at this and see how it works.

File: api/hydrophoneApi.go [hunk 26]
```diff
@@ -510,42 +568,61 @@ func (a *Api) sendModelAsResWithStatus(res http.ResponseWriter, model interface{
...
+			fields = append(fields, field)
+		} else if extraErrs, ok := extra.([]error); ok {
+			if len(extraErrs) > 0 {
+				errs = append(errs, extraErrs...)
+			}
+		} else {
+			nonErrs = append(nonErrs, extra)
+		}
+	}
+	return nonErrs, errs, fields
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/hydrophoneApi.go:625]: Done.

File: api/patientInvites.go (hunk 19 above)
Comment (ewollesen) [reply api/patientInvites.go:103]: Thank you, good catch.

File: api/forgot.go [hunk 27]
```diff
@@ -85,37 +86,27 @@ func (a *Api) passwordReset(res http.ResponseWriter, req *http.Request, vars map
...
 	//unless no email was given we say its all good
 	res.WriteHeader(http.StatusOK)
 }
 
 // find the reset confirmation if it exists and hasn't expired
-func (a *Api) findResetConfirmation(conf *models.Confirmation, ctx context.Context, res http.ResponseWriter) *models.Confirmation {
-
-	log.Printf("findResetConfirmation: finding [%v]", conf)
-	found, err := a.findExistingConfirmation(ctx, conf, res)
+func (a *Api) findResetConfirmation(ctx context.Context, conf *models.Confirmation) (*models.Confirmation, bool, error) {
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/forgot.go:96]: It was only called in once place, and that one place basically duplicated all the logic anyway. I've removed this function.

File: api/forgot.go [hunk 28]
```diff
@@ -133,46 +124,50 @@ func (a *Api) findResetConfirmation(conf *models.Confirmation, ctx context.Conte
...
-	if conf := a.findResetConfirmation(resetCnf, req.Context(), res); conf != nil {
-		if resetCnf.Key == "" || resetCnf.Email != conf.Email {
-			statusErr := &status.StatusError{Status: status.NewStatus(http.StatusBadRequest, STATUS_RESET_ERROR)}
-			a.sendModelAsResWithStatus(res, statusErr, http.StatusBadRequest)
+	conf, expired, err := a.findResetConfirmation(ctx, resetCnf)
+	if err != nil {
+		a.sendError(ctx, res, http.StatusInternalServerError, STATUS_ERR_FINDING_CONFIRMATION, err)
+		return
+	}
+	if expired {
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply api/forgot.go:143]: Yep, that's where I landed basically.

**Comment by ewollesen** [comment]
I believe I've address all of your comments to date, Todd. If you see something that I didn't respond to, call it out, maybe GitHub just hid it from me. :/ But I didn't ignore anything intentionally.

File: api/clinicianInvites.go [hunk 29]
```diff
@@ -394,11 +394,11 @@ func (a *Api) sendClinicianConfirmation(req *http.Request, confirmation *models.
 	}
 
 	if !a.createAndSendNotification(req, confirmation, emailContent) {
-		return http.StatusInternalServerError, STATUS_ERR_SENDING_EMAIL, nil
+		return STATUS_ERR_SENDING_EMAIL, nil
```
Comment (toddkazakov) [inline api/clinicianInvites.go:397]: You should always return an error if you return a message otherwise in some places the code may not return a response, e.g.:
```go

		msg, err := a.sendClinicianConfirmation(req, confirmation)
		if err != nil {
			a.sendError(ctx, res, http.StatusInternalServerError, msg, err)
			return
		}
```

**Review by toddkazakov — CHANGES_REQUESTED** [review]
